### DISCUSSION TITLE
Multiple APIKeyHeader behavior and OpenAPI Specs

### DISCUSSION THREAD

**Discussion by aalifadv** [fastapi/fastapi#9012] [Questions]
### First check

* [X] I added a very descriptive title to this issue.
* [X] I used the GitHub search to find a similar issue and didn't find it.
* [X] I searched the FastAPI documentation, with the integrated search.
* [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [X] I already read and followed all the tutorial in the docs and didn't find an answer.
* [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [X] After submitting this, I commit to implement a Pull Request for a confirmed bug.

Hello!

We can not thank you enough for FastAPI. We truly appreciate all the effort. Thank you millions! :heart_eyes: :heart_eyes:. 
Below, is a description of what we observed -- we are not sure if this is a bug, feature request, lack of documentation or something else. The least we can do is to contribute a fix and we will but we thought to pass this by you first to validate assumption and get guidance. 
 

**What is the problem we are trying to solve:**
We are developing API based on FastAPI, we need to authenticate the client and to do that, we need to get the value of  X-APP-ID & X-APP-KEY from the headers by using fastapi.security.APIKeyHeader. Every client should pass both headers to be considered authenticated.
 
### Example

**First Attempt**

 <details><summary>
Here's a self-contained example (click to expand):
</summary>

```Python
from fastapi import FastAPI, Depends
from fastapi.security import APIKeyHeader

app = FastAPI()

__app_id_scheme = APIKeyHeader(name='X-APP-ID')
__app_key_scheme = APIKeyHeader(name='X-APP-KEY')


@app.get("/")
def test_request(api_id: str = Depends(__app_id_scheme),
                 api_key: str = Depends(__app_key_scheme)):
    return {
        'api_key': api_key,
        'api_id': api_id
    }


security_scheme_paths = app.openapi().get('paths').get("/").get('get').get('security')
security_scheme_components = app.openapi().get('components').get('securitySchemes')

# security_scheme_paths:
# [{'APIKeyHeader': []}, {'APIKeyHeader': []}]
# security_scheme_components:
# {'APIKeyHeader': {'type': 'apiKey', 'in': 'header', 'name': 'X-APP-KEY'}}
```
</details>

### Observation

- run the above, and access the documentation url, e.g http://127.0.0.1:8081/docs
- you will notice the Authentication lock, 
- when clicking on it, the pop up just asks you for one of the keys, which is X-APP-KEY. 

**We were expecting** to see two input fields, one for each APIKeyHeader. 

### Analysis
By examining the app.openapi() ( as shown in the code above),  we notice that the value of security_scheme_paths which is the Security Requirement Object has the two APIKeyHeader with the same name. The security_scheme_components which is the Security Scheme Object has only one APIKeyHeader defined. We think the root cause for not having the Authentication popup asks for both inputs, is that there is only one APIKeyHeader in the Security Scheme Object. Looking at the code in fastapi/openapi/utils.py [line 80](https://github.com/tiangolo/fastapi/blob/33be5fc8baa02bae949e1986f1821dae8f84f487/fastapi/openapi/utils.py#L80), we think that the value of security_name will be APIKeyHeader and hence, the second APIKeyHeader will overwrite the previous one. So we will only get one. 

**Second Attempt**

 <details><summary>
Here's a self-contained example of our second attempt (click to expand):
</summary>

Using the example above, the value of security_name [variable](https://github.com/tiangolo/fastapi/blob/33be5fc8baa02bae949e1986f1821dae8f84f487/fastapi/openapi/utils.py#L79)  is coming from scheme_name in the APIKeyHeader and it is configurable:

```Python
from fastapi import FastAPI, Depends
from fastapi.security import APIKeyHeader

app = FastAPI()

__app_id_scheme = APIKeyHeader(name='X-APP-ID', scheme_name='APIIdHeader')
__app_key_scheme = APIKeyHeader(name='X-APP-KEY', scheme_name='APIKeyHeader')

@app.get("/")
def test_request(api_key: str = Depends(__app_id_scheme),
                 api_name: str = Depends(__app_key_scheme)):
    return {
        'api_key': api_key,
        'api_name': api_name
    }

security_scheme_paths = app.openapi().get('paths').get("/").get('get').get('security')
security_scheme_components = app.openapi().get('components').get('securitySchemes')

# security_scheme_paths:
# [{'APIIdHeader': []}, {'APIKeyHeader': []}]
# # security_scheme_components:
# {'APIIdHeader': {'type': 'apiKey', 'in': 'header', 'name': 'X-APP-ID'}, 'APIKeyHeader': {'type': 'apiKey', 'in': 'header', 'name': 'X-APP-KEY'}}
```

 </details>

### Observation

- run the above, and access the documentation url e.g http://127.0.0.1:8081/docs
- you will notice the Authentication lock,
- when clicking on it, the pop up asks you for both keys and shows two input fields 

**This is what we were expecting**, two input fields, one for each. Also, examining the app.openapi() we notice that two security requirement object and the two elements in the security scheme object. 

_**However**_, it seems that the openapi ui will authenticate using only one of them. For example, if you click on the Authentication Lock and provide only the X-APP-ID, the ui will consider you authenticated. It seems that it accept either of them to consider you authenticated. This is not the desired behavior with openapi ui -- we wanted to be authenticated when both are passed. 

Looking at the [specification](https://swagger.io/specification/#security-requirement-object), we found this:

> Security Requirement Objects that contain multiple schemes require that all schemes MUST be satisfied for a request to be authorized. This enables support for scenarios where multiple query parameters or HTTP headers are required to convey security information.
> 

Also, looking at the [specification](https://swagger.io/docs/specification/authentication/api-keys/#multiple), we found out that multiple api key is supported and there is a way to do AND and OR:

> That is, security is an array of hashmaps, where each hashmap contains one or more named security schemes. Items in a hashmap are combined using logical AND, and array items are combined using logical OR. Security schemes combined via OR are alternatives – any one can be used in the given context. Security schemes combined via AND must be used simultaneously in the same request

So in our example, we need to convert the value of security_scheme_paths  from list of dict, to a list of one dict with multiple APIKeyHeader, from

`[{'APIIdHeader': []}, {'APIKeyHeader': []}]`

to:
`[{'APIIdHeader': [], 'APIKeyHeader': []}]`

**Third Attempt**

<details><summary>
Here's a self-contained example of our third attempt (click to expand):
</summary>

```python
from fastapi import FastAPI, Depends
from fastapi.security import APIKeyHeader
from fastapi.openapi.utils import get_openapi

app = FastAPI()

__app_id_scheme = APIKeyHeader(name='X-APP-ID', scheme_name='APIIdHeader')
__app_key_scheme = APIKeyHeader(name='X-APP-KEY', scheme_name='APIKeyHeader')



@app.get("/")
def test_request(api_key: str = Depends(__app_id_scheme),
                 api_name: str = Depends(__app_key_scheme)):
    return {
        'api_key': api_key,
        'api_name': api_name
    }


def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(
        title="Custom title",
        version="3.0.0",
        description="This is a very custom OpenAPI schema",
        routes=app.routes,
    )
    get_current_sec_req_list = openapi_schema.get('paths').get("/").get('get').get('security')
    list_of_one_dict = [{}]
    for sec_req in get_current_sec_req_list:
        list_of_one_dict[0].update(sec_req)

    openapi_schema['paths']["/"]['get']['security'] = list_of_one_dict
    app.openapi_schema = openapi_schema

    return app.openapi_schema

app.openapi = custom_openapi
```

</details>

### Observation

- run the above, and access the documentation url e.g http://127.0.0.1:8081/docs
- you will notice the Authentication lock,
- when clicking on it, the pop up asks you for both keys and shows two input fields 
- if you only make an input in one field, the lock will continue to be gray.
- one both required fields are filled, the lock be locked and the color will be black.

### Asks

0) Did we over look a better way to accomplish what we wanted ( multiple apikeyheader used for authentication)?
1) Would you be receptive if we submit a PR for better documentation around the usage of APIKeyHeader? 
2) What is your guidance to implement a better support for multiple APIKeyHeader? we are thinking that a suitable solution could be to refactor the function get_openapi_security_definitions in openapi/utils.py so it is aware of multiple APIKeyHeader. It can default to AND (all passed keys are required ) as per the specification above. If you are okay with the direction, we will work on PR.

### Environment

* **OS: [e.g. Linux]:** Linux
* **FastAPI Version [e.g. 0.3.0]:** 0.63.0
* **Python Version:** Python 3.8.5

Thank you,

**Comment by tzmara** [comment]
Long time ago, but did you find a better solution?

**Comment by eugenio-r-tmax** [comment]
I have the same issue here. I have different parts of the API that should respond to different types of API keys, but when using the FastAPI handler (APIKeyHeader), they all get overwritten, and only one is handled.

**Comment by BilalAlpaslan** [comment]
i think this pr solve this problem. https://github.com/tiangolo/fastapi/pull/4791

**Reply by eugenio-r-tmax** [reply]
I tried your code exactly like this.

```
from fastapi import FastAPI, Security
from fastapi.security import api_key

app = FastAPI()

api_key_header_a = api_key.APIKeyHeader(name='key_a')
api_key_header_b = api_key.APIKeyHeader(name='key_b')


@app.get("/a")
async def a(auth_token: str = Security(api_key_header_a)):
    pass


@app.get('/b')
async def b(auth_token: str = Security(api_key_header_b)):
    pass
```

swagger show only `key_b` for both ep

l'm on fastapi  0.104.1

**Reply by BilalAlpaslan** [reply]
No, this pr will be fix this bug, but you can turn around the bug like this for now: https://github.com/tiangolo/fastapi/issues/4635

**Reply by BilalAlpaslan** [reply]
```py
from fastapi import FastAPI, Security
from fastapi.security import api_key

app = FastAPI()

api_key_header_a = api_key.APIKeyHeader(name='key_a', scheme_name='api_key_header_a')
api_key_header_b = api_key.APIKeyHeader(name='key_b', scheme_name='api_key_header_b')


@app.get("/a")
async def a(auth_token: str = Security(api_key_header_a, use_cache=False)):
    pass


@app.get('/b')
async def b(auth_token: str = Security(api_key_header_b, use_cache=False)):
    pass
```

**Reply by eugenio-r-tmax** [reply]
Now it works perfectly, thank you very much :pray:

**Reply by AndreWohnsland** [reply]
Late to the party, but thank you for this solution. Exactly my problem and what I needed to solve it :)
